src/zal.rs

```diff
@@ -0,0 +1,185 @@
+//! This module provides "ZK Acceleration Layer" traits
+//! to abstract away the execution engine for performance-critical primitives.
+//!
+//! The ZAL Engine is voluntarily left unconstrained
+//! so that accelerator libraries are not prematurely limited.
+//!
+//! Terminology
+//! -----------
+//!
+//! We use the name Backend+Engine for concrete implementations of ZalEngine.
+//! For exaple H2cEngine for pure Halo2curves implementation.
+//!
+//! Alternative names considered were Executor or Driver however
+//! - executor is already used in Rust (and the name is long)
+//! - driver will be confusing as we work quite low-level with GPUs and FPGAs.
+//!
+//! Unfortunately Engine is used in bn256 for pairings.
+//! Fortunately ZalEngine is only used in the prover
+//! while "pairing engine" is only used in the verifier
+//!
+//! Initialization design space
+//! ---------------------------
+//!
+//! It is recommended that ZAL backends provide:
+//! - an initialization function:
+//!   - either "fn new() -> ZalEngine" for simple libraries
+//!   - or a builder pattern for complex initializations
+//! - a shutdown function.
+//!
+//! The ZalEngine can be a stub type
+//! and the shutdown function might be unnecessary
+//! if the ZalEngine uses a global threadpool like Rayon.
+//!
+//! Backends might want to add as an option:
+//! - The number of threads (CPU)
+//! - The device(s) to run on (multi-sockets machines, multi-GPUs machines, ...)
+//! - The curve (JIT-compiled backend)
+
+use crate::msm::best_multiexp;
+use pasta_curves::arithmetic::CurveAffine;
+
+// The ZK Accel Layer API
+// ---------------------------------------------------
+
+pub trait ZalEngine {}
+
+pub trait MsmAccel<C: CurveAffine>: ZalEngine {
+    fn msm(&self, coeffs: &[C::Scalar], base: &[C]) -> C::Curve;
```

Comment:
The user of this trait can either use this method, or the cached version.  Perhaps this method could have a default implementation that uses the cached version underneath?